### PR TITLE
changes vmagent api

### PR DESCRIPTION
**Pull request by f41gh7**
adds auth.Token to promremotewrite InsertHandlerReader
changes remoteWrite client constructor, allows to use multiple remoteWriteUrl schemes, like kafka://
changes url path concatenation for tenant remoteWrite

### CODE REVIEW

File: app/vmagent/remotewrite/client.go [hunk 1]
```diff
@@ -84,43 +86,51 @@ type client struct {
 	stopCh chan struct{}
 }
 
-func newClient(argIdx int, remoteWriteURL, sanitizedURL string, fq *persistentqueue.FastQueue, concurrency int) *client {
```
Comment (valyala) [inline app/vmagent/remotewrite/client.go:87]: How about just renaming this function to `newHTTPClient` and not modifying it contents except of setting `c.sendBlock = c.sendBlockHTTP`? Then the caller could determine which client to create from the remoteWriteURL scheme and call the corresponding function. E.g. if the scheme is `foobar`, then the caller could call `newFoobarClient()`?

**Review by valyala — CHANGES_REQUESTED** [review]

File: app/vmagent/remotewrite/remotewrite.go [hunk 2]
```diff
@@ -181,17 +183,21 @@ func newRemoteWriteCtxs(at *auth.Token, urls []string) []*remoteWriteCtx {
...
+	for i, remoteWriteURLRaw := range urls {
+		remoteWriteURL, err := url.Parse(remoteWriteURLRaw)
+		if err != nil {
+			logger.Fatalf("invalid remoteWriteURL: %q, err: %s", remoteWriteURL, err)
+		}
 		sanitizedURL := fmt.Sprintf("%d:secret-url", i+1)
 		if at != nil {
 			// Construct full remote_write url for the given tenant according to https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#url-format
-			remoteWriteURL = fmt.Sprintf("%s/insert/%d:%d/prometheus/api/v1/write", remoteWriteURL, at.AccountID, at.ProjectID)
+			remoteWriteURL.Path = path.Join(remoteWriteURL.Path, fmt.Sprintf("/insert/%d:%d/prometheus/api/v1/write", at.AccountID, at.ProjectID))
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply app/vmagent/remotewrite/remotewrite.go:194]: RemoteWriteURL may contain some query params - `extra_label=key=value` or any other. In this case url will be broken.

File: app/vmagent/remotewrite/client.go (hunk 1 above)
Comment (f41gh7) [reply app/vmagent/remotewrite/client.go:87]: Sounds good to me!

File: app/vmagent/remotewrite/remotewrite.go (hunk 2 above)
Comment (valyala) [inline app/vmagent/remotewrite/remotewrite.go:194]:
```suggestion
			remoteWriteURL.Path = fmt.Sprintf("%s/insert/%d:%d/prometheus/api/v1/write", remoteWriteURL.Path, at.AccountID, at.ProjectID)
```

File: app/vmagent/remotewrite/remotewrite.go [hunk 3]
```diff
@@ -3,6 +3,8 @@ package remotewrite
 import (
 	"flag"
 	"fmt"
+	"net/url"
+	"path"
```
Comment (valyala) [inline app/vmagent/remotewrite/remotewrite.go:7]:
```suggestion
```

**Review by valyala — APPROVED** [review]

**Comment by valyala** [comment]
Thanks!

**Pull request merged**